### PR TITLE
Fix numerical issue in NUTS acceptance ratio computation

### PR DESCRIPTION
**Pull request by junpenglao**
Should reduce the frequency of getting "Mass matrix contains zeros on the diagonal" during warmup and fix a bunch of issue in https://github.com/pymc-devs/pymc3/issues/3959

### CODE REVIEW

File: pymc3/step_methods/hmc/nuts.py [hunk 1]
```diff
@@ -384,23 +390,24 @@ def _build_subtree(self, left, depth, epsilon):
...
     def stats(self):
         # Update accept stat if any subtrees were accepted
         if self.log_size > 0:
-            # Remove contribution from initial state which is always a perfect 
-	    # accept
-            sum_weight = np.expm1(self.log_size)
-            self.mean_tree_accept = np.exp(self.log_accept_sum) / sum_weight
+            # Remove contribution from initial state which is always a perfect
+            # accept
+            log_sum_weight = logdiffexp_numpy(self.log_size, 0.0)
```
Comment (aseyboldt) [inline pymc3/step_methods/hmc/nuts.py:412]: Why not use `log1mexp_numpy` directly?

**Comment by michaelosthege** [comment]
Please don't merge yet - two notebook re-runs are still active (GP-TProcess and ODE_API_Introduction).

I will push all successful re-runs in the morning.

File: pymc3/math.py [hunk 2]
```diff
@@ -136,19 +176,28 @@ def logsumexp(x, axis=None):
...
+
 def invlogit(x, eps=sys.float_info.epsilon):
     """The inverse of the logit function, 1 / (1 + exp(-x))."""
-    return (1. - 2. * eps) / (1. + tt.exp(-x)) + eps
+    return (1.0 - 2.0 * eps) / (1.0 + tt.exp(-x)) + eps
+
+
+def logbern(log_p):
+    if np.isnan(log_p):
+        raise FloatingPointError("log_p can't be nan.")
```
Comment (ColCarroll) [inline pymc3/math.py:199]: wow, who knew `FloatingPointError` is a builtin?

**Review by ColCarroll — APPROVED** [review]
this looks great -- nice fix!

**Review by aseyboldt — COMMENTED** [review]
Seems that the PR fixes the issue, I'm not 100% on what *exactly* it actually was though.
I'm still a bit confused about why we get a nan, and not just an inf. In the old code:
`sum_weight` is `expm1(logsumexp([0, *stuff]))`, so always 0, positive or +inf.
`exp(log_accept_sum)` is `exp(logsumexp(stuff))`, where `stuff` is an array of elements of [0, 1]. So it can be 0 or positive. +inf can't happen, because the probabilities we sum up can't be >1.
So we end up with a nan, iff both are 0. Is that what's going on?

File: pymc3/math.py (hunk 2 above)
Comment (michaelosthege) [reply pymc3/math.py:199]: If only pm.exceptions.ShapeError was a builtin too...

**Comment by michaelosthege** [comment]
These notebooks were fixed:

+ [x] `Euler-Maruyama_and_SDEs` Mass matrix contains zeros on the diagonal
+ [x] `GLM-hierarchical` Mass matrix contains zeros on the diagonal
+ [x] `GLM-rolling-regression` Mass matrix contains zeros on the diagonal
+ [x] `GP-SparseApprox` Mass matrix contains zeros on the diagonal
+ [x] `censored_data` Mass matrix contains zeros on the diagonal
+ [x] `probabilistic_matrix_factorization` Mass matrix contains zeros on the diagonal
+ [x] `putting_workflow` Mass matrix contains zeros on the diagonal
+ [x] `stochastic_volatility` Mass matrix contains zeros on the diagonal.

**Comment by junpenglao** [comment]
@aseyboldt we are getting nan because both `np.exp(self.log_accept_sum)` and `sum_weight` overflow after exponential, and inf/inf -> nan

File: pymc3/step_methods/hmc/nuts.py (hunk 1 above)
Comment (junpenglao) [reply pymc3/step_methods/hmc/nuts.py:412]: logdiffexp_numpy calls log1mexp_numpy internally anyway, I think using logdiffexp is a bit clear but i dont have strong preference.

**Comment by aseyboldt** [comment]
Ah, I was confused because of this here:
https://github.com/pymc-devs/pymc3/pull/3656/files#diff-36963bcefab91495fd993259e82d9b53R331
I assumed each individual log_p_accept would be <0.
I see now.
Maybe we should call it self.log_weighted_accept_sum then? It really isn't a prob anymore, right?

**Comment by junpenglao** [comment]
Yeah good idea, done.

File: pymc3/step_methods/hmc/nuts.py (hunk 1 above)
Comment (aseyboldt) [reply pymc3/step_methods/hmc/nuts.py:412]: I think it is a bit cleaner to use the more specialized function that doesn't do the (small) extra work that's unnecessary here. But both really is fine

**Pull request merged**